cpp/src/io/comp/comp.cpp

```diff
@@ -98,8 +119,128 @@ std::vector<std::uint8_t> compress_snappy(host_span<uint8_t const> src,
   return cudf::detail::make_std_vector_sync<uint8_t>(d_dst, stream);
 }
 
+void device_compress(compression_type compression,
+                     device_span<device_span<uint8_t const> const> inputs,
+                     device_span<device_span<uint8_t> const> outputs,
+                     device_span<compression_result> results,
+                     rmm::cuda_stream_view stream)
+{
+  if (compression == compression_type::NONE) { return; }
+
+  auto const nvcomp_type = to_nvcomp_compression(compression);
+  auto nvcomp_disabled   = nvcomp_type.has_value() ? nvcomp::is_compression_disabled(*nvcomp_type)
+                                                   : "invalid compression type";
+  if (not nvcomp_disabled) {
+    return nvcomp::batched_compress(*nvcomp_type, inputs, outputs, results, stream);
+  }
+
+  switch (compression) {
+    case compression_type::SNAPPY: return gpu_snap(inputs, outputs, results, stream);
+    default: CUDF_FAIL("Compression error: " + nvcomp_disabled.value());
+  }
+}
+
+void host_compress(compression_type compression,
+                   device_span<device_span<uint8_t const> const> inputs,
+                   device_span<device_span<uint8_t> const> outputs,
+                   device_span<compression_result> results,
+                   rmm::cuda_stream_view stream)
+{
+  if (compression == compression_type::NONE) { return; }
+
+  auto const num_chunks = inputs.size();
+  auto h_results        = cudf::detail::make_host_vector<compression_result>(num_chunks, stream);
+  auto const h_inputs   = cudf::detail::make_host_vector_async(inputs, stream);
+  auto const h_outputs  = cudf::detail::make_host_vector_async(outputs, stream);
+  stream.synchronize();
+
+  std::vector<std::future<size_t>> tasks;
+  auto const streams = cudf::detail::fork_streams(stream, h_comp_pool().get_thread_count());
```

Comment:
Nit: should we cap the stream pool size to 32 here to avoid the warning in `get_streams`? (https://github.com/rapidsai/cudf/blob/dc2a75cba40d38f4a6ba66e652764e96fa6b593d/cpp/src/utilities/stream_pool.cpp#L131)

Reply:
I forgot about that warning! Yup, should be capped at stream pool size.

Reply:
We can consider moving the `STREAM_POOL_SIZE` variable to the header `stream_pool.hpp` and then invoke `fork_streams` as
```suggestion
  auto const streams = cudf::detail::fork_streams(stream, std::min(STREAM_POOL_SIZE, h_comp_pool().get_thread_count()));
```

Reply:
I'm trying to use  `cudf::detail::global_cuda_stream_pool().get_stream_pool_size()`, that should work without changes to the pool

Reply:
Ooh, good idea about the thread count env var, I'll use it in the JSON PR #17708 as well!